### PR TITLE
Add VIME Qwen3-8B TP4/CP2 consistency experiment and results

### PR DESCRIPTION
**Pull request by inaniloquentee**
## What this PR adds

- Adds the Qwen3-8B 200-step GRPO consistency example for one 8×H100 node.
- Integrates RL-Kernel deterministic attention, dense FFN, collectives, and linear logp with the VIME/Megatron training path.
- Adds strict fail-closed provider/runtime checks, CUDA Graph contract checks, per-step train/rollout logp validation, snapshot validation, aggregation, and reproducible plotting.
- Optimizes the strict TP4/CP2 path with CP communication staging/reuse, deterministic collective updates, LM-head/logit reuse for linear logp, Megatron/VIME runtime adapters, and a Qwen FFN path.
- Publishes the completed, matched-workload G10 production P/P and optimized G11 RL-Kernel R/R comparison: all 200 paired steps, summaries, bootstrap intervals, and figures.

## Completed experiment scope

| Group | Attention / FFN / logp route | Ray submission | Result |
|---|---|---|---|
| G10 | production / production (P/P) | `vime200-g10-n200-b8-s16-refkl001-s1234-20260902a` | `SUCCEEDED`, 200/200 |
| Optimized G11 | RL-Kernel / RL-Kernel (R/R), strict consistency | `vime200-g11-point19-n200-b8-s16-refkl001-s1234-20260903a` | `SUCCEEDED`, 200/200 |

Both runs used the same workload configuration. The identity fields and implementation route are intentionally different.

| Item | Matched value |
|---|---|
| Hardware | 1 node, 8× NVIDIA H100 80GB |
| Actor parallelism | TP4 / CP2 / PP1 |
| Rollout engines | 2 engines, TP4 each |
| Horizon | 200 rollout/training steps |
| Seeds | training `1234`, rollout `1234` |
| Sampling | 8 prompts × 16 samples = global batch 128 |
| Response limit | 7,168 tokens |
| Dynamic batching | maximum 4,096 tokens/GPU |
| vLLM memory utilization | 0.4 |
| KL loss | enabled, coefficient `0.001` |
| Snapshot requirement | exactly 8 rank files per completed step |

## Version provenance

| Arm | RL-Kernel | VIME | Notes |
|---|---|---|---|
| G10 | `b3bfd2b` | `1a113710` | Existing production P/P baseline |
| Optimized G11 | [`6a7646c`](https://github.com/RL-Align/RL-Kernel/commit/6a7646c0737eda3b3cee101d38be1e583d13931d) | [`41a95b3`](https://github.com/RL-Align/vime/commit/41a95b3867f3901b2675b586d03d1f1b8e45e8a4) | Strict RL-Kernel R/R implementation measured here |
| Result bundle | [`40db4d3`](https://github.com/RL-Align/RL-Kernel/commit/40db4d31982cd4a7ba28fbc96982b2af1f62921d) | — | 200-step CSV, JSON, Markdown, and plots; formatting-only follow-up [`c663a71`](https://github.com/RL-Align/RL-Kernel/commit/c663a71652e0452deb535c7b81bd46b63b46d643) |

This is a matched-workload implementation comparison, not a claim that both arms ran the same implementation commit. G10 was not rerun because this update changes the strict RL-Kernel G11 route, not the production P/P route.

## Full 200-step performance result

| Metric (arithmetic mean) | G10 P/P | Optimized G11 R/R | G11 relative result |
|---|---:|---:|---:|
| Rollout time | 130.22 s/step | 82.75 s/step | **36.5% faster** |
| Rollout throughput | 672.39 tok/GPU/s | 1,134.00 tok/GPU/s | **68.7% higher** |
| Longest-sample throughput | 55.51 tok/s | 87.62 tok/s | **57.8% higher** |
| Reference logp time | 20.90 s/step | 20.92 s/step | 0.1% slower; bootstrap CI includes parity |
| Actor train time | 80.51 s/step | 107.18 s/step | 33.1% slower |
| Actor train throughput | 8,993.65 tok/s | 7,050.92 tok/s | 21.6% lower |
| Total train time | 101.65 s/step | 128.33 s/step | 26.3% slower |
| End-to-end step time | 251.99 s/step | 231.27 s/step | **8.2% faster** |
| Mean response length | 5,492.25 tokens | 5,757.01 tokens | G11 generated 4.8% more tokens |
| Mean raw reward | 0.528555 | 0.491445 | G10−G11 = +0.037109 |
| Mean KL loss | 0.056965 | 0.003783 | reported, not treated as a speed metric |

The paired end-to-end step-time effect is **+8.2% in favor of G11**, with a 95% paired-bootstrap interval of **[+7.2%, +9.2%]**. Reference logp is statistically at parity in this run (**−0.1%**, 95% interval **[−1.0%, +0.8%]**).

### Performance matrix

G10 versus optimized G11 200-step performance matrix

### Paired performance statistics

G10 versus optimized G11 paired performance effects with 95 percent bootstrap intervals

### Stage summary and 200-step trajectories

G10 versus optimized G11 performance summary

G10 versus optimized G11 performance trajectories

## Strict train/rollout consistency and reward

- Optimized G11: `mismatch_count == 0` and `max_abs_diff == 0` at every one of the 200 steps.
- G10 production P/P mismatch count per step: n=200, zero rate 0%, min 1,121.12, median 2,040.76, p90 3,661.65, p95 3,858.31, p99 4,342.79, max 4,758.01.
- G10 maximum absolute logp difference per step: n=200, min 0.602204, median 0.978281, p90 1.285081, p95 1.315587, p99 1.453950, max 1.591547.
- Paired mean reward difference (G10−G11): +0.037109, with 95% bootstrap CI [+0.030664, +0.043633]. This is a paired-step interval from one training seed, not a multi-seed generalization interval.
- Reference KL (`rollout/kl`) is exactly zero for both arms in the recorded logs; KL-loss values above come from `train/kl_loss`.

G10 versus optimized G11 reward, KL loss, and consistency trajectories

### Mean train/rollout logp difference

G10 versus optimized G11 mean absolute train/rollout logp difference

## Data integrity and method

- Both Ray submissions finished successfully with 200/200 steps.
- Both runs have 1,600 train-data snapshot files: exactly 8 expected ranks for every step, with no missing, duplicate, or unexpected rank.
- All metrics above use 200 strictly aligned step pairs. Required values had to exist and be finite; no imputation or row deletion was allowed.
- Percentage intervals use a paired non-parametric bootstrap over aligned steps with fixed seed `1234` and 20,000 resamples.
- `rollout tokens/GPU/s = mean response tokens × 128 samples / (rollout seconds × 8 GPUs)`.
- `longest-sample tokens/s = maximum response tokens / rollout seconds`.
- A vLLM EngineCore error appeared only during teardown after all data was complete; Ray still reported `SUCCEEDED`. No fatal training error or traceback preceded completion.

## Reproducible artifacts

- [Result README](https://github.com/RL-Align/RL-Kernel/tree/40db4d31982cd4a7ba28fbc96982b2af1f62921d/examples/vime_qwen3_8b_tp4_cp2_200/results/scale_reference_s1234_g10_g11_optimized)
- [All 200 paired steps (CSV)](https://github.com/RL-Align/RL-Kernel/blob/40db4d31982cd4a7ba28fbc96982b2af1f62921d/examples/vime_qwen3_8b_tp4_cp2_200/results/scale_reference_s1234_g10_g11_optimized/rounds.csv)
- [Distribution and bootstrap summary (JSON)](https://github.com/RL-Align/RL-Kernel/blob/40db4d31982cd4a7ba28fbc96982b2af1f62921d/examples/vime_qwen3_8b_tp4_cp2_200/results/scale_reference_s1234_g10_g11_optimized/summary.json)
- [Regeneration script](https://github.com/RL-Align/RL-Kernel/blob/c663a71652e0452deb535c7b81bd46b63b46d643/examples/vime_qwen3_8b_tp4_cp2_200/results/scale_reference_s1234_g10_g11_optimized/plot_report.py)
- [VIME companion branch](https://github.com/RL-Align/vime/tree/codex/g11-tp4-cp2-linear-logp-reuse)

## Validation

Targeted integration/operator suite: **183 passed**.

```text
tests/test_framework_operator_integrations.py
tests/test_framework_runtime_adapters.py
tests/test_linear_logp.py
tests/test_vime_linear_logp_provider.py
tests/test_qwen_ffn.py
tests/test_cp_attention.py
tests/test_vime_tp4_example.py
```

The completed 200-step optimized G11 run is also the end-to-end distributed validation for the code captured in `6a7646c`.

### CODE REVIEW

**Review by Flink-ddd — CHANGES_REQUESTED** [review]
Please regenerate the overlapping text in the images and add a mismatch count comparison between the vime baseline and RL-Kernel.

File: examples/vime_qwen3_8b_tp4_cp2_200/validate_run.py [hunk 1]
```diff
@@ -0,0 +1,501 @@
...
+            records = [
+                value["operators"][module]
+                for value in matching
+                if isinstance(value.get("operators"), Mapping)
+                and module in value["operators"]
+            ]
+            installed_count = sum(
+                module in value.get("installed_hooks", {}) for value in matching
+            )
+            call_count = sum(int(record.get("call_count", 0)) for record in records)
```
Comment (leonardHONG) [inline examples/vime_qwen3_8b_tp4_cp2_200/validate_run.py:170]: The current aggregation can pass with just one reporter. Should we also check that we got the full expected set of ranks and engines here?

File: examples/vime_qwen3_8b_tp4_cp2_200/validate_run.py [hunk 2]
```diff
@@ -0,0 +1,501 @@
...
+        ),
+        "engine_mode": bool(re.search(r"cudagraph_mode.*FULL_DECODE_ONLY", log_text)),
+        "not_eager": "enforce_eager=False" in log_text,
+        "capture_sizes": (
+            f"'cudagraph_capture_sizes': {capture_sizes}" in log_text
+            or f'"cudagraph_capture_sizes":{compact_sizes}' in log_text
+        ),
+    }
+    return {
+        "passed": all(checks.values()),
```
Comment (leonardHONG) [inline examples/vime_qwen3_8b_tp4_cp2_200/validate_run.py:102]: I think this only verifies the configuration string. Do we have a way to check that both vLLM engines actually captured and replayed the graph?

File: examples/vime_qwen3_8b_tp4_cp2_200/validate_run.py [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+        report = {
+            "schema_version": "rlkernel.vime_qwen3_8b_tp4_cp2_200.validation.v1",
+            "passed": False,
+            "errors": [f"{type(exc).__name__}: {exc}"],
+        }
+    output.write_text(
+        json.dumps(report, indent=2, sort_keys=True) + "\n", encoding="utf-8"
+    )
+    print(json.dumps(report, indent=2, sort_keys=True))
+    if args.seal and report["passed"]:
```
Comment (leonardHONG) [inline examples/vime_qwen3_8b_tp4_cp2_200/validate_run.py:495]: Could we wait for a persisted Ray terminal status before marking this complete? Otherwise, a failure after the last logged step could still get sealed as successful.

File: csrc/cuda/distributed/deterministic_collective.cu [hunk 4]
```diff
@@ -198,8 +205,153 @@ __global__ void deterministic_all_reduce_fast_kernel(
   }
   __syncthreads();
 
-  for (int64_t index = threadIdx.x; index < element_count; index += blockDim.x) {
-    output[index] = fixed_tree_reduce<T, WorldSize>(peers, index);
+  if constexpr (std::is_same_v<T, nv_bfloat16>) {
+#if (__CUDA_ARCH__ >= 800 || !defined(__CUDA_ARCH__))
+    const int64_t pair_count = element_count / 2;
+    auto* pair_output = reinterpret_cast<nv_bfloat162*>(output);
```
Comment (leonardHONG) [inline csrc/cuda/distributed/deterministic_collective.cu:211]: contiguous() doesn't necessarily give us the alignment required by nv_bfloat162. Could we check the pointer alignment here and fall back to scalar bf16 when it's not aligned?

**Comment by inaniloquentee** [comment]
## Follow-up: version-matched 200-step G10/G11 scale-reference run

This is a new matched comparison, separate from the previously published sealed runs. Both arms use the same clean revisions and workload:

- RL-Kernel `b3bfd2beef97a00d9f9aac678f5d251504fbfa85`
- VIME `1a113710e80aa7cfc271caa9bd90bcf348a7af08`
- Megatron-LM `1dcf0dafa884ad52ffb243625717a3471643e087`
- 1 node / 8×H100 80GB, TP4/CP2, 200 steps, seed/rollout seed 1234
- rollout batch 8 prompts × 16 samples = global batch 128, max response 7,168, max tokens/GPU 4,096
- rollout-logp reuse enabled; reference KL loss enabled with coefficient 0.001

Both Ray submissions succeeded at 200/200. Each has 1,600 train-data snapshots, exactly 8 ranks for every step. The manifest audit found no unexpected common-configuration differences; expected differences are the G10 P/P versus G11 R/R routes, strict provider/CI gates, and identity-derived fields/paths.

### Paired results, steps 1–200

| Metric | G10 mean | G11 mean | G10−G11 | n |
|---|---:|---:|---:|---:|
| Raw reward | 0.5286 | 0.4914 | +0.0371 | 200 |
| Reference KL loss | 0.0570 | 0.0038 | +0.0532 | 200 |
| Rollout time (s) | 130.22 | 82.40 | +47.82 | 200 |
| Rollout tokens/GPU/s | 672.39 | 1,138.13 | −465.75 | 200 |
| Longest-sample tokens/s | 55.51 | 87.94 | −32.43 | 200 |
| Reference log-prob time (s) | 20.90 | 30.78 | −9.87 | 200 |
| Actor train time (s) | 80.51 | 131.67 | −51.16 | 200 |
| Actor train tokens/s | 8,993.65 | 5,736.97 | +3,256.68 | 200 |
| Total train time (s) | 101.65 | 162.68 | −61.03 | 200 |
| Total step time (s) | 251.99 | 265.23 | −13.25 | 200 |

G11 improves rollout throughput by 69.3% and reduces rollout time by 36.7%. G10 is 56.8% faster in actor-train throughput and has 38.9% lower actor-train time, so the end-to-end result favors G10 by 13.25 s/step (5.0%) on this run.

The paired reward difference is **+0.037109 G10−G11**, with a paired-step 95% bootstrap CI of **[+0.030586, +0.043555]** (seed 1234, 20,000 resamples). This interval describes variation across the 200 paired steps of this single-seed run; it is not a multi-seed confidence interval.

### Train/rollout log-probability consistency

- **G11:** all 200 steps are strictly bitwise clean: `mismatch_count == 0` and `max_abs_diff == 0` at every step (0 mismatches over an estimated 147,379,363 active response tokens).
- **G10:** all 200 steps have nonzero mismatch, as expected for the production P/P comparison. Per-step `mismatch_count`: median 2,040.76, p95 3,858.31, max 4,758.01. Per-step `max_abs_diff`: median 0.9783, p95 1.3156, max 1.5915 at step 158. Estimated token-level agreement is 58.58% (58,230,217 mismatches over 140,601,694 active response tokens).

Metric definitions: rollout tokens/GPU/s = mean response tokens × 128 / (rollout seconds × 8 GPUs); longest-sample throughput = maximum response tokens / rollout seconds. All required fields were finite for all 200 paired steps; no imputation or row deletion was used. The logged rollout-stage `rollout/kl` was zero in both arms, while the reference-model `train/kl_loss` is reported separately above.

The attached archive contains the full 200-step CSV and the reproducible Markdown summary. SHA-256: `2dc82f953201f349b410f47099e2e7ebe5d9b0cd991f651cc74b21b81bb727bd`.
[g10_g11_version_matched_results.zip](https://github.com/user-attachments/files/31768415/g10_g11_version_matched_results.zip)

**Review by maxiaosong1124 — APPROVED** [review]

**Review by Flink-ddd — APPROVED** [review]
LGTM
